### PR TITLE
feat!: Add support for permissions composed using DRF

### PR DESCRIPTION
**Pull request by xitij2000**
**Description:**
Currently rest_condition is used to compose permission classes, however support
for composition is now part of DRF since v3.9 and rest_condition has not seen
updates in a while.

This change adds support for DRF-style composition to the middleware.

https://github.com/edx/upgrades/issues/2

**Reviewers:**
- [ ] 

**Merge checklist:**
- [ ] All reviewers approved
- [x] CI build is green
- [x] Version bump if needed
- [x] Changelog record added
- [x] Documentation updated (not only docstrings)
- [x] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is
      finished.
- [ ] Delete working branch (if not needed anymore)

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @xitij2000! I've created [OSPR-6026](https://openedx.atlassian.net/browse/OSPR-6026) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

File: edx_rest_framework_extensions/auth/jwt/middleware.py
Comment (meysam81) [inline edx_rest_framework_extensions/auth/jwt/middleware.py:47]: There's a typo here:

```suggestion
        Given a composed permission class or instance, return its component parts.
```

File: edx_rest_framework_extensions/auth/jwt/tests/test_middleware.py
Comment (meysam81) [inline edx_rest_framework_extensions/auth/jwt/tests/test_middleware.py:187]: This comment is the same as the one before. I think it should be modified accordingly.

**Review by meysam81 — CHANGES_REQUESTED** [review]
I think the **Additional Details** of the merge request needs more description cause it's from the template and hence makes it uneasy to track what's going on here.

**Comment by xitij2000** [comment]
@meysam81  Yup. this PR is a WIP since the PR it's based on it still WIP and not merged. I will update it when the groundwork for that PR is complete.

**Comment by natabene** [comment]
@xitij2000 Thank you for your contribution. Please let me know once this is ready for our review.

File: edx_rest_framework_extensions/auth/jwt/middleware.py
Comment (xitij2000) [inline edx_rest_framework_extensions/auth/jwt/middleware.py:62]: This code is moved to an iterator above.

File: edx_rest_framework_extensions/auth/jwt/middleware.py
Comment (xitij2000) [inline edx_rest_framework_extensions/auth/jwt/middleware.py:65]: We can probably remove this and rest_condition, but I'm to sure if both should be supported for a while.

File: requirements/base.in
Comment (xitij2000) [inline requirements/base.in:13]: Remove rest condition from base requirements.

File: requirements/test.in
Comment (xitij2000) [inline requirements/test.in:13]: Add it in test requirements so that tests against it can still be run. If we can completely drop it then even better.

**Comment by xitij2000** [comment]
@meysam81  Could you have a look and approve if all looks good?

File: CHANGELOG.rst
Comment (meysam81) [inline CHANGELOG.rst:21]: This sentence has two `work` in it.
Also, the structure is a little bit ambiguous.

File: edx_rest_framework_extensions/auth/jwt/middleware.py
Comment (meysam81) [inline edx_rest_framework_extensions/auth/jwt/middleware.py:65]: I suggest we raise a warning here to inform the use of a "deprecated" -ish library, hinting to be upgraded in a later release.

**Review by meysam81 — COMMENTED** [review]
How can I test the result of this change in a local dev environment?

File: edx_rest_framework_extensions/auth/jwt/middleware.py
Comment (xitij2000) [reply edx_rest_framework_extensions/auth/jwt/middleware.py:65]: I think on balance this code can just be removed, since using this release already means switching to a version of django that rest-condition doesn't support.

Comment (xitij2000) [reply edx_rest_framework_extensions/auth/jwt/middleware.py]: Removed.

File: requirements/base.in
Comment (xitij2000) [reply requirements/base.in:13]: A lot of packages use rest-condition via this package, so this should resolve it for a lot of other packages.

File: docs/index.rst
Comment (meysam81) [inline docs/index.rst:14]: @xitij2000 
Can we be sure that this upgrade won't break any dependent library? I'm just thinking that this is a breaking change and might need a thorough testing in all the environments (3.5, 3.6, 3.8).

The following two places can be changed accordingly:

https://github.com/edx/edx-drf-extensions/blob/17a7e337b1aa142fff336dd5fca68399b129b9ba/tox.ini#L2

https://github.com/edx/edx-drf-extensions/blob/17a7e337b1aa142fff336dd5fca68399b129b9ba/.github/workflows/ci.yml#L16

**Review by meysam81 — CHANGES_REQUESTED** [review]
@xitij2000
All looks good to me; just make sure the change of the python versions is well-thought and well-tested. Otherwise, it's ready to be merged.

File: docs/index.rst
Comment (xitij2000) [reply docs/index.rst:14]: I just removed the versions that I knew were definitely not supported anymore. I will reduce it to just the ones tested by the CI.

**Review by meysam81 — APPROVED** [review]
This PR looks good to me. Well done @xitij2000 

- [ ] I tested this
- [x] I read through the code
- [ ] I checked for accessibility issues
- [x] Includes documentation
- [ ] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository.

**Comment by xitij2000** [comment]
> @xitij2000 Thank you for your contribution. Please let me know once this is ready for our review.

This is ready for review.

File: edx_rest_framework_extensions/auth/jwt/middleware.py
Comment (robrap) [inline edx_rest_framework_extensions/auth/jwt/middleware.py:51]: This comment could be a `#` comment next to moved code. See later comment about moving `decomposed_permissions` code.

File: edx_rest_framework_extensions/auth/jwt/middleware.py
Comment (robrap) [inline edx_rest_framework_extensions/auth/jwt/middleware.py:56]: I wonder if it really is ok to skip negations. If so, maybe we need to better document why.
1. Can the negation be of a composed permission? Can there be double negations?
2. Since we aren't really checking for sound boolean logic, but just making sure a particular permission is somehow included, I wonder if we could inadvertently be causing issues by skipping negations. Or, does this end up matching earlier logic? Do we have a test around this from before/after?

File: edx_rest_framework_extensions/__init__.py
Comment (robrap) [inline edx_rest_framework_extensions/__init__.py:3]: 1. I've been wondering whether or not this needs to be considered backward-incompatible. We'll see.
2. On a related compatibility note, someone should understand https://github.com/encode/django-rest-framework/issues/6500 and whether we need any tests around this, or whether it will cause any compatibility changes, etc.

File: edx_rest_framework_extensions/auth/jwt/middleware.py
Comment (robrap) [inline edx_rest_framework_extensions/auth/jwt/middleware.py:61]: Might be useful to have a docstring in addition to any detailed comments.

File: edx_rest_framework_extensions/auth/jwt/middleware.py
Comment (robrap) [inline edx_rest_framework_extensions/auth/jwt/middleware.py:63]: Not sure if this comment about `rest_condition` is current or not, but I have some related questions.
1. Composed conditions may have been created outside of edx-drf-extensions that will pass through this middleware, right? Will this middleware need to complain about that? Will it work?
2. Will `rest_condition` still need to be a test dependency for a while to ensure we do whatever we decide we should be doing in this case?
3. Based on our decisions, is this a backward incompatible change that requires a major version bump?

**Review by robrap — COMMENTED** [review]
Thank you @xitij2000 for getting this started. In addition to my other comments:
1. Can you ensure there is an edx-platform PR that tests against your updated package? I know it includes its own tests of some of this functionality.
2. If the edx-platform tests pass, I wonder if you could try [removing this test override](https://github.com/edx/edx-platform/blob/9bd85dc929d0186b6f9dade1ba23d01ab29263cb/lms/djangoapps/certificates/apis/v0/tests/test_views.py#L281-L285) to see if they continue to pass? Note to self: this is related to ARCHBOM-543. I don't think you can see this ticket.
Thank you!

File: edx_rest_framework_extensions/auth/jwt/middleware.py
Comment (xitij2000) [reply edx_rest_framework_extensions/auth/jwt/middleware.py:71]: Yes, initially I was trying to support both rest condition and DRF, both use differnt mechanisms to get these classes so moving this out to a separate method kept things cleaner. I agree that now it's no longer helping. Will remove.

File: edx_rest_framework_extensions/auth/jwt/middleware.py
Comment (xitij2000) [reply edx_rest_framework_extensions/auth/jwt/middleware.py:56]: The reason I'm not including negation is that this middleware is about making sure that the `NotJwtRestrictedApplication` permission is required for all view classes that use JWT. If the `NotJwtRestrictedApplication` class is applied in a negation, then it is explicitly not applied since the permission is checking for the absence of that permission. 

I think, taking a step back and looking at the reason for this middleware, though, it might be OK to count negations. The reason being that any place that is explicitly negating this permission while using JWT is already aware of this permission and is explicitly crating a set of permissions based on that.

As for inadvertently causing issues. rest_condition and composed permissions in general are used very sparingly in the platform, and I have a [PR ](https://github.com/edx/edx-platform/pull/28663) that touches that code so we can see already that such a case isn't currently present in the platform. 

It may affect plugins, but then the move from rest condition to DRF will be breaking in that case anyway.

File: edx_rest_framework_extensions/auth/jwt/middleware.py
Comment (xitij2000) [reply edx_rest_framework_extensions/auth/jwt/middleware.py:63]: It is not current. For a while I kept compatibility with both, and included it as a test dependency, however I realised that since rest condition isn't compatible with django 3+ and I was removing the usage from the platform, keeping the test around didn't make much sense. 

It is a backwards incompatible change, since without handling rest condition, the platform with just crash if it uses rest condition or uses any plugins that use it.

**Review by robrap — COMMENTED** [review]
Thanks! This is basically ready, with some minor issues to wrap up.

I recommend we do the following:
1. Add contraint to edx-platform now for edx-drf-extensions<8.0.0 with a comment explaining we will remove it in https://github.com/edx/edx-platform/pull/28663. We can land and merge that. It is added protection to ensure that this won't get upgraded outside of your PR. It should be a very quick two-line change. Just tag me on the PR.
2. Update https://github.com/edx/edx-platform/pull/28663 to remove the edx-drf-extensions constraint and to use this latest 8.0.0 update.
3. Merge this any time after edx-platform is constrained and this is approved.

File: requirements/common_constraints.txt
Comment (xitij2000) [reply requirements/common_constraints.txt:22]: I just ran make upgrade and committed the results. I'm not very familiar with the new mechanism for common constraints.

Comment (awais786) [reply requirements/common_constraints.txt:22]: These changes are fine. We have removed following items from global constraint file so it is removing those from local common constraints also.

File: edx_rest_framework_extensions/__init__.py
Comment (robrap) [reply edx_rest_framework_extensions/__init__.py]: @xitij2000: Can you take a quick look at the issue I linked to and see if you think we need any tests, or comments, or code changes, or nothing. While your head is in this work. Thank you!

File: requirements/common_constraints.txt
Comment (robrap) [reply requirements/common_constraints.txt:22]: Thank you both. I wasn’t aware these would be kept in sync. Marking as resolved.

File: edx_rest_framework_extensions/__init__.py
Comment (robrap) [reply edx_rest_framework_extensions/__init__.py]: Note: I get that this may be independent, so if you want to punt, that is ok.

**Review by robrap — APPROVED** [review]
I’d like the edx-platform constraint before merging, but feel free to squash. Thank you!

File: edx_rest_framework_extensions/__init__.py
Comment (xitij2000) [reply edx_rest_framework_extensions/__init__.py]: I did look into the issue, but forgot to update here. 

First, from what I understand, this very same issue exists for "rest_condition" as well. 

The issue in brief is that if you're using object-level permissions, but don't have object-level permissions defined, they default to true from the base class. 

Here is a case where that can be unexpected. 
1. We have a permission `ReadOnly`  that only allows read operation. Here, we've implemented `has_permission`  to be True if the `request.method` is safe method, and didn't implement `has_object_permission` so it's always true (by default).
2. We have a permission like `IsUserCourseAdmin`. Here, `has_permission` can probably return True if the user is a global admin and `has_object_permission` can be true if the user has the right privileges. 

If the above two are compose as `ReadOnly | IsUserCourseAdmin` you'd expect that the read only requests would be allowed for all users, and write requests would be allowed for course admins. 

However, the above permission would **always** resolve to true for all courses. The reason being that our `ReadOnly` permission class **didn't** override the default behaviour of always returning True for `has_object_permission`. 

I think this just means we should be careful when composing permission that apply to objects and make sure that the classing we're composing all specify the behaviour correctly. In the above example the issue could be avoided if you also implemented `has_object_permission` for read only with the same behaviour as `has_permission`.

Comment (robrap) [reply edx_rest_framework_extensions/__init__.py]: Thanks for this. It is very helpful. 
1. Since we only have a small handful of composed permissions that you are touching across the two PRs (this and edx-platform), do you know if they are all safe to be composed?
2. For `ReadOnly`, although it would be a backward incompatible change, does it make sense to override with comment like was done in https://github.com/encode/django-rest-framework/issues/6500#issuecomment-630175431? Note: I'm just asking the question. I know this change is outside of the scope of this PR.

**Comment by robrap** [comment]
@xitij2000: When you get the chance, can you confirm if you are ok with the plan in https://github.com/edx/edx-drf-extensions/pull/202#pullrequestreview-768102458? Thank you.

File: edx_rest_framework_extensions/__init__.py
Comment (xitij2000) [reply edx_rest_framework_extensions/__init__.py]: I don't think we have any 'ReadOnly' permission class. Generally very few places use object-level permissions. 

I think I had checked all the permission we've composed in https://github.com/edx/edx-platform/pull/28663 and none of them use object-level permissions. 

I've checked again they don't.

**Comment by xitij2000** [comment]
> @xitij2000: When you get the chance, can you confirm if you are ok with the plan in [#202 (review)](https://github.com/edx/edx-drf-extensions/pull/202#pullrequestreview-768102458)? Thank you.

Oh, yes! I think I forgot to share this PR: https://github.com/edx/edx-platform/pull/28893
Sorry about that.

File: edx_rest_framework_extensions/__init__.py
Comment (robrap) [reply edx_rest_framework_extensions/__init__.py]: Thanks for checking.

**Comment by robrap** [comment]
@xitij2000: I forgot about the commit message. Would you mind adding a `!` to the title and adding a `BREAKING CHANGE:` comment to the description? See https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html.  Thank you!

**Comment by xitij2000** [comment]
@robrap  I've updated the commit message.

**Comment by robrap** [comment]
Thanks @xitij2000. The OEP and my comment called for two changes to the commit message. You can just copy the BREAKING CHANGE message from the changelog to the commit message, but drop the rst bold formatting to match the OEP. Thank you.

**Comment by xitij2000** [comment]
> Thanks @xitij2000. The OEP and my comment called for two changes to the commit message. You can just copy the BREAKING CHANGE message from the changelog to the commit message, but drop the rst bold formatting to match the OEP. Thank you.

Sorry, I clearly haven't had enough coffee yet. While copying that message though I realised that it isn't quite accurate. 

Support for rest_condition is dropped, which means that if you have permissions using rest_condition, the middleware will throw an error and the platform won't start. I've updated the message accordingly.

**Comment by edx-webhook** [comment]
@xitij2000 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-drf-extensions&entry.867055334=https://github.com/edx/edx-drf-extensions/pull/202&entry.1484655318=https://github.com/xitij2000&entry.752974735=2021-09-07+08:36&entry.1917517419=2021-10-04+05:41&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**
